[autoscaler] Adding Azure Support

gramhagen opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

## Why are these changes needed?
Adding autoscaler for Azure

## Related issue number
Closes #1438 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

AmplabJenkins commented:
Can one of the admins verify this patch?

gramhagen commented:
Remaining items

- [ ] Test AzureNodeProvider functionality
- [ ] Resolve FIXMEs in AzureNodeProvider
- [ ] Address remaining TODOs
- [ ] Update documentation on required packages / setup

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/21528/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/21755/
Test FAILed.

gramhagen commented:
I think this is in reasonable shape. @richardliaw is there a standard example that can be used for an end-to-end test?

richardliaw commented:
Nothing that runs in CI, but maybe take the Ray Tune Distributed
Experiments guide and use the example training script and cluster yaml?

Let me know if you need pointers!

On Tue, Feb 11, 2020 at 4:53 PM Scott Graham <notifications@github.com>
wrote:

> I think this is in reasonable shape. @richardliaw
> <https://github.com/richardliaw> is there a standard example that can be
> used for an end-to-end test?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/7080?email_source=notifications&email_token=ABCRZZOB76LNHBFGO46AHCDRCNCARA5CNFSM4KREF4HKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOELO263A#issuecomment-584953708>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCRZZM6RU6L2XSVTBEHENDRCNCARANCNFSM4KREF4HA>
> .
>

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/21774/
Test FAILed.

richardliaw commented on python/ray/autoscaler/azure/example-gpu-docker.yaml:
```diff
@@ -0,0 +1,136 @@
+            publisher: microsoft-dsvm
+            offer: linux-data-science-vm-ubuntu
+            sku: linuxdsvmubuntu
+            version: latest
```
Is it possible to pin the version of this image?

gramhagen replied:
will do

richardliaw commented on python/ray/autoscaler/azure/example-gpu-docker.yaml:
```diff
@@ -0,0 +1,136 @@
+            offer: linux-data-science-vm-ubuntu
+            sku: linuxdsvmubuntu
+            version: latest
+    # Additional properties in the azure docs.
```
Can you provide a commented out example of another property we can add here? For example, like pre-emptible instances or disk size?

gramhagen replied:
added example example-full.yaml

richardliaw commented on python/ray/autoscaler/azure/example-gpu-docker.yaml:
```diff
@@ -0,0 +1,136 @@
+
+# Custom commands that will be run on the head node after common setup.
+head_setup_commands:
+    - pip install boto3==1.4.8  # 1.4.8 adds InstanceMarketOptions
```
Do we need boto for azure?

gramhagen replied:
nope, will remove

richardliaw commented on python/ray/autoscaler/azure/example-full.yaml:
```diff
@@ -0,0 +1,158 @@
+provider:
+    type: azure
+    location: westus2
+    service_principal: http://ray-service
```
Hm, what is this? Does this work for any user starting to use the azure autoscaler?

gramhagen replied:
I'm using a service principal to restrict control within azure to the resource group being used. in the event credentials are leaked the damage will be contained and the specific service principal can be decommissioned. config.py handles creation of the service principal which should work with any azure user's account (unless they are restricted from creating them, though it's unlikely they would have authority to create the other resources needed).

gramhagen replied:
on second thought we're switching to use managed service identities to remove the pain of creating service principals. will leave it up to the user if they wish to set up service principals to manage security / access.

richardliaw commented on python/ray/autoscaler/azure/example-full.yaml:
```diff
@@ -0,0 +1,158 @@
+            publisher: microsoft-dsvm
+            offer: linux-data-science-vm-ubuntu
+            sku: linuxdsvmubuntu
+            version: latest
```
Can we pin to a specific version of the data science VM?

gramhagen replied:
yep, will do

eisber commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,351 @@
+
+def _configure_resource_group(config):
+    # TODO: look at availability sets
+    # https://docs.microsoft.com/en-us/azure/virtual-machines/windows/tutorial-availability-sets
```
Availability Sets make sure the VMs you deploy on Azure are distributed across multiple, isolated hardware nodes, in a cluster.

eisber commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,351 @@
+from azure.mgmt.authorization import AuthorizationManagementClient
+from azure.mgmt.network import NetworkManagementClient
+from azure.mgmt.resource import ResourceManagementClient
+import paramiko
```
Used for key generation

gramhagen replied:
is there a preferred way to generate the ssh keys?

gramhagen commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,351 @@
+            }
+        }
+    }
+}
```
yes, that sounds cleaner

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22132/
Test FAILed.

gramhagen commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,351 @@
+    resource_client = get_client_from_cli_profile(ResourceManagementClient)
+    auth_client = get_client_from_cli_profile(AuthorizationManagementClient)
+
+    sp_name = config["provider"]["service_principal"]
```
done

gramhagen commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,351 @@
+        app_name = sp_name
+        sp_name = "http://" + sp_name
+    else:
+        app_name = sp_name.split("://", 1)[-1]
```
yep got it

gramhagen commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,351 @@
+                                                           for c in password)
+                    and any(c.isdigit() for c in password)
+                    and any(not c.isalnum() for c in password)):
+                break
```
yeah, i'm beginning to think this will be easier

gramhagen replied:
switching to managed service identities (msi)

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22243/
Test FAILed.

gramhagen commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,351 @@
+    logger.info("Using subscription id: %s", subscription_id)
+    config["provider"]["subscription_id"] = subscription_id
+
+    resource_group = config["provider"]["resource_group"]
```
yep, will do

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22255/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22256/
Test PASSed.

gramhagen commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,351 @@
+        node_config = DEFAULT_NODE_CONFIG.copy()
+        node_config.update(config.get(node_type, {}))
+        config[node_type] = node_config
+    return config
```
yep, got it

gramhagen commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,351 @@
+        with open(auth_path, "w") as f:
+            json.dump(credentials, f)
+
+    config["provider"]["auth_path"] = auth_path
```
this has been removed with move to managed service identity

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22280/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22297/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22306/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22307/
Test FAILed.

gramhagen commented:
i think this is in good shape now, @eisber and I have been doing some testing. @richardliaw let me know if there are remaining issues to address or who else should review

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22326/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22329/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22498/
Test FAILed.

richardliaw commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,259 @@
+            public_key = f.read()
+    else:
+        public_key, private_key_path = _generate_ssh_keys(key_name)
+        logger.info("SSH key pair created: %s", key_name)
```
@edoakes, @ericl Can one of you suggest a better alternative for this? Right now, we create our own keypair via boto for AWS. 

This PR adds ssh key gen via `paramiko` (installed with Azure).

richardliaw commented:
BTW, @eisber and @gramhagen, thanks a bunch for opening this PR!

gramhagen commented:
@richardliaw on the SkuNotAvailable error
we may have chosen a sku that hasn't been released in all locations yet?
try this to get a list of skus available for the location you're using (assuming westus2)
```
az vm image list --publisher microsoft-dsvm --offer ubuntu-1804 --location westus2
```
Then you can change the publisher, offer, sku, version in the config to match what's available. (might be enough just to change sku from 1804-gen2 -> 1804)

gramhagen commented on python/ray/autoscaler/azure/node_provider.py:
```diff
@@ -0,0 +1,288 @@
+        self.cached_nodes[node_id]["tags"] = node_tags
+
+    def terminate_node(self, node_id):
+        """Terminates the specified node."""
```
sure

gramhagen commented on python/setup.py:
```diff
@@ -176,6 +177,8 @@ def find_version(*filepath):
     "pytest",
     "pyyaml",
+    "jsonschema",
+    "importlib_resources",
```
yeah maybe we can handle these issues in the json schema pr. sorry for mixing them. if we want to remove them from here first I can do that.

eisber replied:
done

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/22544/
Test PASSed.

eisber commented:
> 2 --all -

@gramhagen should we print the command in the error message or even try to list directly?

eisber commented on python/ray/ray-schema.json:
```diff
@@ -0,0 +1,231 @@
+	"$schema": "http://json-schema.org/draft-07/schema#",
+	"$id": "http://github.com/ray-project/ray/python/ray-schema.json",
+	"title": "Ray AutoScalar",
+	"description": "Ray autoscalar schema",
```
done

eisber commented on python/ray/ray-schema.json:
```diff
@@ -0,0 +1,231 @@
+{
+	"$schema": "http://json-schema.org/draft-07/schema#",
+	"$id": "http://github.com/ray-project/ray/python/ray-schema.json",
```
done

richardliaw commented:
OK, got it to work. I had to do the following:

```diff
@@ -73,7 +73,7 @@ auth:
 #       the styles of parameter values are not changed
 head_node:
     hardware_profile:
-        vm_size: Standard_D2s_v3
+        vm_size: Standard_F2s_v2
     storage_profile:
         os_disk:
             create_option: FromImage
Stage this hunk [y,n,q,a,d,K,j,J,g,/,e,?]? n
@@ -82,7 +82,7 @@ head_node:
             # List images https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage
             publisher: microsoft-dsvm
             offer: ubuntu-1804
-            sku: 1804-gen2
+            sku: 1804
             version: 20.02.01
```

Would it be ok to set those as defaults?

BTW, why does node shutdown take a very long time?

eisber commented:
> OK, got it to work. I had to do the following:
> 
> ```diff
> @@ -73,7 +73,7 @@ auth:
>  #       the styles of parameter values are not changed
>  head_node:
>      hardware_profile:
> -        vm_size: Standard_D2s_v3
> +        vm_size: Standard_F2s_v2
>      storage_profile:
>          os_disk:
>              create_option: FromImage
> Stage this hunk [y,n,q,a,d,K,j,J,g,/,e,?]? n
> @@ -82,7 +82,7 @@ head_node:
>              # List images https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage
>              publisher: microsoft-dsvm
>              offer: ubuntu-1804
> -            sku: 1804-gen2
> +            sku: 1804
>              version: 20.02.01
> ```
> 
> Would it be ok to set those as defaults?
> 
> BTW, why does node shutdown take a very long time?

Which VM skus are available is subscription and region dependent. In general we should choose a sku that matches the expected use-case the most (e.g. F machines focus more on CPU than memory).

gramhagen commented:
> > BTW, why does node shutdown take a very long time?

right now I'm blocking on the vm being removed (which take a bit of time). the main reason I was doing that is because an error will pop up if you try to delete a public ip address that is associated with a vm. I think a better way would be to only wait to dissociate the ip then allow for asynchronous deletion of all the resources. I can test out that change.

in the future I'd like to support re-using vms that were terminated, which would mean all the resources would remain, but the vm would be put into a inactive state until the cluster is deleted and everything is removed.

richardliaw commented on python/ray/autoscaler/azure/config.py:
```diff
@@ -0,0 +1,262 @@
+            public_key = f.read()
+    else:
+        public_key, private_key_path = _generate_ssh_keys(key_name)
+        logger.info("Creating SSH key pair: %s", key_name)
```
Is there a reason why we need to generate the ssh keys automatically? 

We do it for boto/aws because each instance needs to be associated with a particular registered keypair. Because that doesn't seem to be the case, I think you should just raise an informative/actionable error message if the private/public key isn't specified

gramhagen replied:
Yeah, I'm fine with having instructions on how to generate the ssh keys and leaving it up to the user before hand. Just wasn't sure what the expectation was here.

richardliaw replied:
OK sounds good!

richardliaw commented:
Left a last comment - @eisber I've also merged your changes into this PR/rebased on master.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/23006/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/23007/
Test FAILed.

eisber commented:
@richardliaw I started rewriting some of our code using ARM templates (see https://github.com/eisber/ray/tree/marcozo/arm):

- no more polling code in python
- reduced dependencies

richardliaw commented:
@eisber awesome! Maybe we can first merge this PR (so that others can play around with it), and then post a followup PR for an ARM-based implementation? 

This PR is very close to merging - we just need to remove the keygen code.

gramhagen commented:
i removed the keygen piece. and updated the configs.
turns out there's no easy way to clean up resources without waiting for the vm to delete, so the long running terminate_node will need to stay as is at least until we move to deallocation vs deletion.

@richardliaw the config files are still pointing to a wheel that I've built for azure. when do new wheels get built? should we push using this custom wheel and then push another pr after the new wheels are available?

@eisber are there remaining changes needed here? other than the arm update?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/23185/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/23217/
Test FAILed.

richardliaw commented:
@gramhagen and @eisber - thanks a lot for your contribution and patience with this PR!

robertnishihara commented:
@gramhagen, this is fantastic, thanks for contributing this!
